src/cli/serve/serve.js

```diff
@@ -305,23 +295,32 @@ export default function (program) {
   }
 
   command.action(async function (opts) {
+    let configs = [getConfigPath(), ...getEnvConfigs(), ...(opts.config || [])];
     const unknownOptions = this.getUnknownOptions();
-    const configs = [getConfigPath(), ...getEnvConfigs(), ...(opts.config || [])];
-    const serverlessMode = getServerlessProjectMode(opts);
-
-    if (serverlessMode) {
-      maybeSetRecentConfig('serverless.recent.yml', serverlessMode, opts.dev, configs, 'push');
-    }
 
     // .dev. configs are "pushed" so that they override all other config files
     if (opts.dev && opts.devConfig !== false) {
-      maybeAddConfig('kibana.dev.yml', configs, 'push');
-      if (serverlessMode) {
-        maybeAddConfig(`serverless.dev.yml`, configs, 'push');
-        maybeAddConfig('serverless.recent.dev.yml', configs, 'push');
+      configs.push(resolveConfig('kibana.dev.yml'));
+    }
+
+    // Filtering out all config paths that didn't exist
+    configs = configs.filter(Boolean);
+
+    const serverlessMode = getServerlessModeFromOpts(opts) || getServerlessModeFromCfg(configs);
```

Comment:
IMO, this check needs to be more complex:

If we keep it like this, it'll break the project switcher's dev functionality. Typical flow:

1. Dev starts Kibana with `yarn start --serverless` (no project specified)
2. The first time, Project Switcher defaults to `es` ✅. Otherwise, it checks the latest project type it was running (it was stored in `serverless.recent.yml`) 🛑 with this approach, `opts.serverless === true` will always return `es`.
3. While running Kibana, the developer could switch the project type. `POST /internal/serverless/switch_project` would write the selected project type in `serverless.recent.yml`, and the watcher would trigger a Kibana restart. `serve.js` reads this new config and starts Kibana in the new project type.

Of course, this logic applies only if the CLI option `--serverless` is `true` (no explicit project selected).

cc @clintandrewhall as the original author of Project Switcher in case I missed any feature 😇

Reply:
Understood! This was an oversight from my side, I automatically placed the concepts together (serverless => project switcher). I'll iron this out in a minute!

Reply:
@afharo - according to the original logic, the most recent `--serverless` choice is also persisted in the file anytime it's a string. I'll adjust my code to reflect this behaviour